### PR TITLE
Fixing `IterativeClosestPointWithNormals::{Const,}Ptr`

### LINKED ISSUE #4416: [icp.h] IterativeClosestPointWithNormals shared_ptr is of type IterativeClosestPoint
**Describe the bug**
I try to pass a shared_ptr of different ICPs arround. With GICP this works like a charm. But with IterativeClosestPointWithNormals the pointer type always falls back to IterativeClosestPoint Class.
Therefore I can not `setUseSymmetricObjective(true)` on that shared_ptr.

**Context**

Pass IterativeClosestPointWithNormals as shared_ptr to a function.

**Expected behavior**

The expected behavior can be seen with GICP which is also based on IterativeClosestPoint but has its own shared_ptr definition.

**Current Behavior**

The IterativeClosestPointWithNormals::Ptr is of type IterativeClosestPoint.

**Screenshots/Code snippets**

```C++
template<typename PointType>
void parametrize(typename pcl::IterativeClosestPointWithNormals<PointType,PointType>::Ptr& icp) {

	std::cout << " class name " << icp->getClassName() << std::endl; // prints IterativeClosestPointWithNormals correctly
	icp->setUseSymmetricObjective(true); // compilation fails because of type IterativeClosestPoint
}

int main() {
pcl::IterativeClosestPointWithNormals<pcl::PointNormal,pcl::PointNormal>::Ptr icp_n (new pcl::IterativeClosestPointWithNormals<pcl::PointNormal,pcl::PointNormal>);
parametrize<pcl::PointNormal>(icp_n);
}
```

**Your Environment (please complete the following information):**

 - OS: [e.g. Ubuntu 20.04]
 - PCL Version [1.11]

**Possible Solution**

In icp.h starting with line 333:

```C++
      using Ptr = shared_ptr<IterativeClosestPoint<PointSource, PointTarget, Scalar> >;
      using ConstPtr = shared_ptr<const IterativeClosestPoint<PointSource, PointTarget, Scalar> >;

```
change to

```C++
      using Ptr = shared_ptr<IterativeClosestPointWithNormals<PointSource, PointTarget, Scalar> >;
      using ConstPtr = shared_ptr<const IterativeClosestPointWithNormals<PointSource, PointTarget, Scalar> >;

```
as GICP does in gicp.h line 99f.:
```C++
      using Ptr = shared_ptr< GeneralizedIterativeClosestPoint<PointSource, PointTarget> >;
      using ConstPtr = shared_ptr< const GeneralizedIterativeClosestPoint<PointSource, PointTarget> >;

```

### PR DESCRIPTION
**Pull request by PerceptMD**
Closes #4416

### CODE REVIEW

File: registration/include/pcl/registration/icp.h [hunk 1]
```diff
@@ -330,8 +330,8 @@ namespace pcl
       using IterativeClosestPoint<PointSource, PointTarget, Scalar>::transformation_estimation_;
       using IterativeClosestPoint<PointSource, PointTarget, Scalar>::correspondence_rejectors_;
 
-      using Ptr = shared_ptr<IterativeClosestPoint<PointSource, PointTarget, Scalar> >;
-      using ConstPtr = shared_ptr<const IterativeClosestPoint<PointSource, PointTarget, Scalar> >;
+      using Ptr = shared_ptr<IterativeClosestPointWithNormals<PointSource, PointTarget, Scalar> >;
```
Comment (kunaltyagi) [inline registration/include/pcl/registration/icp.h:333]: In order to reduce error and set an example for future code, could I ask for a small refactor?

A code in the spirit of the following would be nice:
```cpp
class SomeClass: BaseClass
{
  using Self = SomeClass;
  using Base = BaseClass;
  public:

  using Ptr = shared_ptr<Self>;
  using ConstPtr = shared_ptr<const Self>;

  using Base::something;
};
```

**Review by kunaltyagi — COMMENTED** [review]
LGTM, but perhaps we can set an example for the future

File: registration/include/pcl/registration/icp.h (hunk 1 above)
Comment (PerceptMD) [reply registration/include/pcl/registration/icp.h:333]: Did not see this in any of the ICP classes. So this would be necessary to fix on all of them.

Comment (kunaltyagi) [reply registration/include/pcl/registration/icp.h:333]: Let's start with one class (this PR), and then modify others (followup PRs)

**Comment by PerceptMD** [comment]
What happened? Did Windows Build failed because of my changes?

File: registration/include/pcl/registration/icp.h (hunk 1 above)
Comment (PerceptMD) [reply registration/include/pcl/registration/icp.h:333]: Okay I have committed your suggestions in icp.h.
Should I open up a new pull request or can I add this anywhere here?

Comment (kunaltyagi) [reply registration/include/pcl/registration/icp.h:333]: Thanks. Please create new PRs whenever you can follow up 😄

**Comment by kunaltyagi** [comment]
No, the windows CI is just trigger happy (it runs out of memory)

File: registration/include/pcl/registration/icp.h (hunk 1 above)
Comment (kunaltyagi) [reply registration/include/pcl/registration/icp.h:333]: PS: You need to push the changes, they aren't visible on GitHub

**Comment by larshg** [comment]
Actually this time it was an unit test that failed!

**Comment by PerceptMD** [comment]
I dived into that unit testing and found this log:

> 116: [----------] 1 test from SacTest/1, where TypeParam = class pcl::LeastMedianSquares<struct pcl::PointXYZ>
  116: [ RUN      ] SacTest/1.InfiniteLoop
  116: D:\a\1\s\test\sample_consensus\test_sample_consensus.cpp(133): error: Expected equality of these values:
  116:   std::cv_status::no_timeout
  116:     Which is: 4-byte object <00-00 00-00>
  116:   cv.wait_for (lock, 2s)
  116:     Which is: 4-byte object <01-00 00-00>
  116: [  FAILED  ] SacTest/1.InfiniteLoop, where TypeParam = class pcl::LeastMedianSquares<struct pcl::PointXYZ> (2001 ms)
  116: [----------] 1 test from SacTest/1 (2001 ms total)

This is the code arround that FAILED test
```C++
  // Waits for the delay
  std::unique_lock<std::mutex> lock (mtx);
  #if defined(DEBUG) || defined(_DEBUG)
    EXPECT_EQ (std::cv_status::no_timeout, cv.wait_for (lock, 15s));
  #else
    EXPECT_EQ (std::cv_status::no_timeout, cv.wait_for (lock, 2s)); // <- THIS IS LINE 133!
  #endif
  // release lock to avoid deadlock
  lock.unlock();
  thread.join ();
```

Dont know whats happening here, but it does not seem to be related to my commit.

**Comment by kunaltyagi** [comment]
This failure wasn't related to your code. Why did you close this PR?

**Comment by PerceptMD** [comment]
Sry mate, I pused this into my forked master.
When I created the new PR for UniformSampling, these changes were also visible because of my git branch mess.
So i reset my master to pcl/master.
I did not knew that this would create trouble here, but yes, now I see why.
Idk whats the cleanest way to get things right here. Any suggestions?
If changes are not lost, I can push it from a feature branch of my repo. But I guess this would end in a new PR?

**Comment by kunaltyagi** [comment]
Yeah, pretty much a new branch and a new PR

**Comment by PerceptMD** [comment]
> I dived into that unit testing and found this log:
> 
> > 116: [----------] 1 test from SacTest/1, where TypeParam = class pcl::LeastMedianSquares
> > 116: [ RUN      ] SacTest/1.InfiniteLoop
> > 116: D:\a\1\s\test\sample_consensus\test_sample_consensus.cpp(133): error: Expected equality of these values:
> > 116:   std::cv_status::no_timeout
> > 116:     Which is: 4-byte object <00-00 00-00>
> > 116:   cv.wait_for (lock, 2s)
> > 116:     Which is: 4-byte object <01-00 00-00>
> > 116: [  FAILED  ] SacTest/1.InfiniteLoop, where TypeParam = class pcl::LeastMedianSquares (2001 ms)
> > 116: [----------] 1 test from SacTest/1 (2001 ms total)
> 
> This is the code arround that FAILED test
> 
> ```c++
>   // Waits for the delay
>   std::unique_lock<std::mutex> lock (mtx);
>   #if defined(DEBUG) || defined(_DEBUG)
>     EXPECT_EQ (std::cv_status::no_timeout, cv.wait_for (lock, 15s));
>   #else
>     EXPECT_EQ (std::cv_status::no_timeout, cv.wait_for (lock, 2s)); // <- THIS IS LINE 133!
>   #endif
>   // release lock to avoid deadlock
>   lock.unlock();
>   thread.join ();
> ```
> 
> Dont know whats happening here, but it does not seem to be related to my commit.

Any feedback for this?

**Comment by kunaltyagi** [comment]
@larshg This seems like a timing issue to me. Just running the CI again should pass it, right?

**Comment by larshg** [comment]
Yes, it should. I triee increase the timeout once, but maybe we should try increase it some more, to get it more stable if the cpu power is unstable on the CI :-)

**Pull request closed without merging**